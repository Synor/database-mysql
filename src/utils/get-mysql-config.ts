import { SynorError } from '@synor/core'

type ConnectionConfig = import('mysql').ConnectionConfig

type MySQLDatabaseConfig = Pick<ConnectionConfig, 'ssl'> &
  Required<
    Pick<
      ConnectionConfig,
      'database' | 'host' | 'port' | 'user' | 'password' | 'multipleStatements'
    >
  >

export function getMySQLConfig(uri: string): MySQLDatabaseConfig {
  const {
    protocol,
    pathname,
    hostname,
    port,
    username,
    password,
    searchParams
  } = new URL(uri)

  if (protocol !== 'mysql:') {
    throw new SynorError(`Invalid DatabaseURI`)
  }

  let ssl: ConnectionConfig['ssl']

  if (searchParams.has('ssl')) {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const sslRaw = searchParams.get('ssl')!

    try {
      ssl = JSON.parse(sslRaw)
    } catch {
      ssl = sslRaw
    }
  }

  return {
    database: pathname.replace(/^\//, ''),
    host: hostname,
    port: Number(port),
    user: username,
    password: password,
    multipleStatements: true,
    ssl
  }
}
