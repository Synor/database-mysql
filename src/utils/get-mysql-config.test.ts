import { getMySQLConfig } from './get-mysql-config'

describe('utils:getMySQLConfig', () => {
  let uri: Parameters<typeof getMySQLConfig>[0]

  beforeEach(() => {
    uri = 'mysql://root:root@127.0.0.1:3306/synor'
  })

  test('accepts mysql uri', () => {
    expect(getMySQLConfig(uri)).toMatchSnapshot()
  })

  test(`throws if uri protocol is not 'mysql:'`, () => {
    uri = uri.replace('mysql:', 'postgresql:')
    expect(() => getMySQLConfig(uri)).toThrow()
  })

  test('throws if uri is malformed', () => {
    uri = 'mysql://@_@/synor'
    expect(() => getMySQLConfig(uri)).toThrow()
  })

  test('multipleStatements config is set', () => {
    expect(getMySQLConfig(uri).multipleStatements).toBe(true)
  })

  describe('ssl config', () => {
    beforeEach(() => {
      uri = 'mysql://root:root@127.0.0.1:3306/synor'
    })

    test('accepts string', () => {
      const ssl = 'Amazon RDS'
      uri = `${uri}?ssl=${ssl}`
      expect(getMySQLConfig(uri).ssl).toBe(ssl)
    })

    test('accepts quoted string', () => {
      const ssl = '"Amazon RDS"'
      uri = `${uri}?ssl=${ssl}`
      expect(getMySQLConfig(uri).ssl).toBe(JSON.parse(ssl))
    })

    test('accepts stringified json', () => {
      const ssl = { ca: '/mysql-ca.crt', rejectUnauthorized: false }
      uri = `${uri}?ssl=${JSON.stringify(ssl)}`
      expect(getMySQLConfig(uri).ssl).toMatchObject(ssl)
    })
  })
})
