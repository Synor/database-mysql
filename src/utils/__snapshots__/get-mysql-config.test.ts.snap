// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils:getMySQLConfig accepts mysql uri 1`] = `
Object {
  "database": "synor",
  "host": "127.0.0.1",
  "multipleStatements": true,
  "password": "root",
  "port": 3306,
  "ssl": undefined,
  "user": "root",
}
`;
